### PR TITLE
#359 add empty check for applyBeforeCallback

### PR DESCRIPTION
**Pull request by ahmetoz**
Related bug: #359

### CODE REVIEW

File: src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncOptionsBuilderTest.java [hunk 1]
```diff
@@ -160,6 +164,29 @@ public void applyBeforeUpdateCallBack_WithNullReturnCallback_ShouldReturnEmptyLi
         assertThat(filteredList).isEmpty();
     }
 
+    private abstract class MockTriFunction implements
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncOptionsBuilderTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/363/commits/7005a8498830f07ab7d14fdf9c5e416cff5dde82

File: src/test/java/com/commercetools/sync/types/TypeSyncOptionsBuilderTest.java [hunk 2]
```diff
@@ -158,6 +161,28 @@ public void applyBeforeUpdateCallBack_WithNullReturnCallback_ShouldReturnEmptyLi
         assertThat(filteredList).isEmpty();
     }
 
+    private abstract class MockTriFunction implements
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/types/TypeSyncOptionsBuilderTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/363/commits/7005a8498830f07ab7d14fdf9c5e416cff5dde82

File: src/test/java/com/commercetools/sync/products/ProductSyncOptionsBuilderTest.java [hunk 3]
```diff
@@ -173,10 +177,33 @@ public void applyBeforeUpdateCallBack_WithNullReturnCallback_ShouldReturnEmptyLi
         assertThat(filteredList).isEmpty();
     }
 
+    private abstract class MockTriFunction implements
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/products/ProductSyncOptionsBuilderTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/363/commits/7005a8498830f07ab7d14fdf9c5e416cff5dde82

File: src/test/java/com/commercetools/sync/categories/CategorySyncOptionsBuilderTest.java [hunk 4]
```diff
@@ -174,6 +178,28 @@ public void applyBeforeUpdateCallBack_WithNullReturnCallback_ShouldReturnEmptyLi
         assertThat(filteredList).isEmpty();
     }
 
+    private abstract class MockTriFunction implements
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/categories/CategorySyncOptionsBuilderTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/363/commits/7005a8498830f07ab7d14fdf9c5e416cff5dde82

File: src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java [hunk 5]
```diff
@@ -177,6 +177,11 @@ public int getBatchSize() {
     public List<UpdateAction<U>> applyBeforeUpdateCallBack(@Nonnull final List<UpdateAction<U>> updateActions,
                                                            @Nonnull final V newResourceDraft,
                                                            @Nonnull final U oldResource) {
+
+        if (updateActions.isEmpty()) {
+            return updateActions;
+        }
+
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java]: https://github.com/commercetools/commercetools-sync-java/pull/363/commits/29bdc28647b0b09c26a3f0720bfc62a165bbc689

**Comment by ahmetoz** [comment]
> Please add a release notes entry

Could you check this commit:   https://github.com/commercetools/commercetools-sync-java/pull/363/commits/bedd52f0f4b9f3413d2a423a9d1721e1fb99223f

File: docs/RELEASE_NOTES.md [hunk 6]
```diff
@@ -43,6 +43,7 @@
     - **ProductType Sync** - `ProductTypeSyncUtils#buildActions`, `ProductTypeUpdateActionUtils#buildAttributesUpdateActions`  
     now treat the values of the optional fields `isSearchable`, `inputHint` and `attributeConstraint` 
     as (`true`, `SingleLine` and `None` respectivley) if they are `null` or not passed. [#354](https://github.com/commercetools/commercetools-sync-java/issues/354)
+    - **Commons** - Fixed a bug in the `beforeUpdateCallback` which callback should not be called when there are no update actions. [#359](https://github.com/commercetools/commercetools-sync-java/issues/359)
```
Comment (heshamMassoud) [inline docs/RELEASE_NOTES.md]:
```suggestion
    - **Commons** - Fixed a bug in the `beforeUpdateCallback` which caused the callback to be called even on an empty list of update actions. [#359](https://github.com/commercetools/commercetools-sync-java/issues/359)
```

File: src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java [hunk 7]
```diff
@@ -161,23 +161,26 @@ public int getBatchSize() {
...
+     * and returns the same list. If the result of the callback is null, an empty list is returned.
      *
      * @param updateActions the list of update actions to apply the {@code beforeUpdateCallback} function on.
      * @param newResourceDraft the new resource draft that is being compared to the old resource.
      * @param oldResource the old resource that is being compared to the new draft.
      * @return a list of update actions after applying the {@code beforeUpdateCallback} function on. If the
-     *         {@code beforeUpdateCallback} function was null, the supplied list of {@code updateActions} is returned as
-     *         is. If the return of the callback was null, an empty list is returned.
+     *         {@code beforeUpdateCallback} function was null or {@code updateActions} was empty, the supplied list of
+     *         {@code updateActions} is returned as is. If the return of the callback was null, an empty list is
```
Comment (ahalberkamp) [inline src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java]: ...function is null

File: src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncOptionsBuilderTest.java [hunk 8]
```diff
@@ -160,6 +164,29 @@ public void applyBeforeUpdateCallBack_WithNullReturnCallback_ShouldReturnEmptyLi
...
+        TriFunction<List<UpdateAction<ProductType>>, ProductTypeDraft, ProductType, List<UpdateAction<ProductType>>> {
+    }
+
+    @Test
+    public void applyBeforeUpdateCallBack_WithEmptyUpdateActions_ShouldNotApplyBeforeUpdateCallback() {
+        final MockTriFunction beforeUpdateCallback = mock(MockTriFunction.class);
+
+        final ProductTypeSyncOptions productTypeSyncOptions = ProductTypeSyncOptionsBuilder.of(CTP_CLIENT)
+                                                                                           .beforeUpdateCallback(
+                                                                                               beforeUpdateCallback)
```
Comment (ahalberkamp) [inline src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncOptionsBuilderTest.java]: that formatting looks strange

File: src/test/java/com/commercetools/sync/types/TypeSyncOptionsBuilderTest.java [hunk 9]
```diff
@@ -158,6 +161,28 @@ public void applyBeforeUpdateCallBack_WithNullReturnCallback_ShouldReturnEmptyLi
...
+        TriFunction<List<UpdateAction<Type>>, TypeDraft, Type, List<UpdateAction<Type>>> {
+    }
+
+    @Test
+    public void applyBeforeUpdateCallBack_WithEmptyUpdateActions_ShouldNotApplyBeforeUpdateCallback() {
+        final MockTriFunction beforeUpdateCallback = mock(MockTriFunction.class);
+
+        final TypeSyncOptions typeSyncOptions = TypeSyncOptionsBuilder.of(CTP_CLIENT)
+                                                                      .beforeUpdateCallback(
+                                                                          beforeUpdateCallback)
```
Comment (ahalberkamp) [inline src/test/java/com/commercetools/sync/types/TypeSyncOptionsBuilderTest.java]: here too

File: src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java [hunk 10]
```diff
@@ -161,23 +161,26 @@ public int getBatchSize() {
...
+     *         returned.
      */
     @Nonnull
     public List<UpdateAction<U>> applyBeforeUpdateCallBack(@Nonnull final List<UpdateAction<U>> updateActions,
                                                            @Nonnull final V newResourceDraft,
                                                            @Nonnull final U oldResource) {
+
         return ofNullable(beforeUpdateCallback)
+            .filter(callback -> !updateActions.isEmpty())
             .map(callBack -> emptyIfNull(callBack.apply(updateActions, newResourceDraft, oldResource)))
```
Comment (ahalberkamp) [inline src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java]: I would rename it to filteredCallback,like
`
 return ofNullable(beforeUpdateCallback)
            .filter(callBack -> !updateActions.isEmpty())
            .map(filteredCallBack -> emptyIfNull(filteredCallBack.apply(updateActions, newResourceDraft, oldResource))
`

**Pull request merged**